### DISCUSSION TITLE
help - shell workflows and findings

### DISCUSSION THREAD

**Discussion by ferasdour** [caido/caido#1621] [Q&A]
not really sure how to get a a workflow that uses shell commands to create a finding, I've tried a few different ways, ensured that stdout would be sent, ensured that process returns code 0, not really sure what else to do here. I've attached a couple examples of ones that for me aren't creating findings.

[crawl url.json](https://github.com/user-attachments/files/17356035/crawl.url.json)
[wpscan-docker.json](https://github.com/user-attachments/files/17356036/wpscan-docker.json)

Kind of using this as an orchestration device and I know that may not be the intended goal of the shell feature, but would like to also find out a way to send findings back from shell in the event of things like ssh/ansible into other devices to run commands. Anyway, figure someone has already run into something like this, so if anyone can help me out real quick, would be much appreciated.

**Comment by Sytten** [comment]
The integrations between shell and caido is basically non existent right now.

What I suggest would be to use the JS node with the module `child_process` (similar to node) to launch your bash script or whatever tool.

Then it will be much easier to get back data on the stdout and create the finding from JavaScript.

Not at the computer right now but I will try to create a guide for that.

**Comment by ferasdour** [comment]
yeah, tried to just use create finding in the ui based on the shell, but it never would work. I guess I'd think that it could go start -> shell -> create findings using shell.data as reference -> end. But i do have some examples to go off of using javascript to create the findings, something (roughtly written) like this maybe?:
```
async function run(input, sdk) {
    const {
        request,
        response
    } = input;
    const responseBody = response.getBody()?.toText();
    cont responseHost = response.getHeader(name: Host)
        const {
        exec
    } = require('child_process');
    if (!responseBody) {
        sdk.console.log("response body is empty");
        return;
    }
    exec('podman run --tls-verify=false -it --rm wpscanteam/wpscan --proxy http://172.25.128.1:8080 --random-user-agent --disable-tls-checks --url ${responseHost} --enumerate ap,at,tt,cb,dbe,u,m  -f cli-no-color', (err, stdout, stderr) => {
        if (err) {
            console.error(`Error: ${err}`);
            return;
        }
        const findingDescription = `Response:\n\n ${stdout}`;
        await sdk.findings.create({
            title: "Paths Found",
            reporter: "Linkfinder",
            request: request,
            description: findingDescription,
            severity: "info",
        });
    }
        export {
        run
    };

```

that makes sense, just didn't know if there was some trick to it with shell.

**Comment by Sytten** [comment]
```js
import { spawn } from "child_process";

async function driveChild(child) {
  let output = "";
  child.stdout.on("data", (data) => {
    output += data.toString();
  });

  let error = "";
  child.stderr.on("data", (data) => {
    error += data.toString();
  });

  const exitCode = await new Promise((resolve, reject) => {
    child.on("close", resolve);
  });

  if (exitCode) {
    throw new Error(`subprocess error exit ${exitCode}, ${error}`);
  }

  return output;
}

export async function run() {
  const child = spawn("podman", ["run"]);
  const result = await driveChild(child);
  const output = JSON.parse(result);
  if (output.something) {
    await sdk.findings.create({
            title: "Paths Found",
            reporter: "Linkfinder",
            request: request,
            description: findingDescription,
            severity: "info",
        });
  }
}
```
@ferasdour That is what you need more or less. Your problem was that you are trying to await in a sync callback.

**Comment by ferasdour** [comment]
Thank you!
